lib/portlayer/store/store.go

```diff
@@ -0,0 +1,105 @@
+// Copyright 2016 VMware, Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package store
+
+import (
+	"errors"
+	"fmt"
+	"net/url"
+	"os"
+
+	"github.com/vmware/vic/pkg/kvstore"
+	"github.com/vmware/vic/pkg/trace"
+	"github.com/vmware/vic/pkg/vsphere/datastore"
+	"github.com/vmware/vic/pkg/vsphere/session"
+
+	log "github.com/Sirupsen/logrus"
+
+	"golang.org/x/net/context"
+)
+
+// Stores map holds a reference to each of the created stores
+// TODO: need to add mutex and protect map during creation, etc
+var Stores map[string]*kvstore.KeyValueStore
+
+// StoresDatastoreURL is the ds path for all ds backed stores
+var StoresDatastoreURL url.URL
+
+const (
+	// default store folder name
+	dsFolder = "kvStores"
+	// available via portLayer API
+	APIKV = "apiKV"
+)
+
+var (
+	ErrDuplicateName = errors.New("duplicate store name")
+	ErrInvalidName   = errors.New("invalid store name, must be regexp[] compliant")
+)
+
+// Init will initialize the package vars and create the default portLayerKV store
+func Init(ctx context.Context, session *session.Session, imgStoreURL url.URL) error {
+	defer trace.End(trace.Begin(imgStoreURL.String()))
+
+	// map of stores to allow other modules to create
+	Stores = make(map[string]*kvstore.KeyValueStore)
+
+	// datastore location for ds backed stores
+	StoresDatastoreURL = imgStoreURL
+
+	//create or restore the api accessible datastore backed k/v store
+	return createDatastoreKeyValue(ctx, session, APIKV)
+}
+
+// NewDatastoreKeyValue will validate the supplied name and then call the package
+// level createDatastoreKeyValue for k/v creation
+func NewDatastoreKeyValue(ctx context.Context, session *session.Session, name string) error {
```

Comment:
In this code it doesn't -- provides the ability for others (later commits) to create k/v stores.  In my next commit it will be called via init and the package level createDatastoreKeyValue will go away